### ISSUE TITLE
Segment results always show as having just occurred.

### ISSUE THREAD

**Issue by zoffline** [zoffline/zwift-offline#1]
First approximation: this is due to not implementing GET/PUT on relay/worlds/[id]/attributes. Implement attributes and see what happens.

**Comment by zoffline** [comment]
Investigated a bit further, the issue is because the segment result time is based on the "world time" rather than the real time. As of now I'm unsure exactly what the world time is compared with the real time. Currently, zoffline just multiplies the real time by 1000 for the world time.

The world time does appear to be a multiple of real time (albeit a random multiple of 64.4131403573055), i.e. it's consistently proportional to real time. I'm testing with zoffline's world time set to that multiple, but haven't seen success so far (not much testing yet though).

tl;dr: need to figure out exactly what Zwift's "world time" is and test more with time.time() * 64.4131403573055 as the world time.

**Comment by oldnapalm** [comment]
To test this should I just change
https://github.com/zoffline/zwift-offline/blob/f5c88b9150591fe933faca69cea0284815530fda/zwift_offline.py#L448
to
`result.world_time = int(time.time())*64.4131403573055`
?

Do you know what's f17 in segment_result table? I have 1970-01-13T17:15:16Z (1098916) for every row.

**Comment by zoffline** [comment]
Change that and also:
https://github.com/zoffline/zwift-offline/blob/f5c88b9150591fe933faca69cea0284815530fda/zwift_offline.py#L448

If it has an effect it will only be for newly created segment results. I can't recall if I ever got to testing a new segment result though...  I would assume I tested at least one.

Don't know what f17 is, but it's always that date in traces with the official Zwift server so it's seemingly unused. Not sure why it's close to but not quite the Unix Epoch though...
iirc f17 isn't sent when requesting session results, only when creating one.

**Comment by oldnapalm** [comment]
Would this work for previously created segment results? (it rounds to 4 decimal places)
`UPDATE segment_result SET world_time = world_time/1000*64.4131403573055`

**Comment by zoffline** [comment]
Sorry for the delayed response and thanks for your testing and insight. After further tests and tweaks there's still no fix for this issue. I added a response for the relay/worlds/[id]/attributes you mentioned, but that didn't help. I was hopeful it was due to sending Zwift more SegmentResult fields than it expected, but fixing that didn't help either. I adjusted the world_time to roughly match Zwift's. No real good reason for that, but opted to do it.

Getting real head scratching now since near as I can tell zoffline is behaving the same as the official Zwift server w.r.t. world time and segment results. Next step will be replaying responses from Zwift's official server and slowly replacing them with zoffline responses to narrow down what's not working with segment result timing. It may be awhile before I do that.

**Comment by oldnapalm** [comment]
Don't worry, thanks for your time.

Do you still think this issue is related to relay/worlds/[id]/attributes?

Doesn't Zwift client send data to server with relay/worlds/[id]/attributes POST method?

Doing a request.stream.read() inside that method I see some data, not sure if it's actually from that request or what most part of the data mean. I isolated the text and other bytes that repeat every request, don't know if it's of any help.

```
1001186922 55 10E8071801200128 C99AD2E603            30003A0B  [my name]  80DDE3AEF5 0258 97BF02 600068889805700078 F203 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 5A 10E8071801200128 C99AD2E683 8080808001 30003A0B  [my name]  8087EEA6F5 0258 A0A831 600068889805700078 A701 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 5A 10E8071801200128 D6BEEEFB83 8080808001 30003A0B  [my name]  80FEE888F2 0258 89B540 600068889805700078 D601 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 55 10E8071801200128 E8E6EAF813            30003A0B  [my name]  80C8C0BFF4 0258 CDE101 600068889805700078 D003 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 59 10E8071801200128 E8E6EAF893 8080808001 30003A0B  [my name]  C0978FB2F4 0258 CF71   600068889805700078 E903 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 55 10E8071801200128 B689F2F123            30003A0B  [my name]  80C2C2BFF4 0258 FDB10B 600068889805700078 9C02 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 5A 10E8071801200128 B689F2F1A3 8080808001 30003A0B  [my name]  80BB82EFF0 0258 95EE12 600068889805700078 9D02 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 56 10E8071801200128 F0C092B0D9 11         30003A0B  [my name]  C0D7EDD1F4 0258 BA9B79 600068889805700078 C201 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 59 10E8071801200128 E8B6C5FAC3 EC90BC71   30003A0B  [my name]  809FF1D5F2 0258 CBAB26 600068889805700078 F001 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 5A 10E8071801200128 ABDCB5ED83 8080808001 30003A0B  [my name]  C09A8DD6F6 0258 E18B42 600068889805700078 9901 8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
1001186922 54 10E8071801200128 FDD8FEA808            30003A0B  [my name]  808FC893F6 0258 A9F817 600068889805700078 75   8001018A0114  [1970-01-13T17:15:16Z]  98010048C0A907
```

**Comment by zoffline** [comment]
It's probably not related to attributes. The client does POST the segment results it just received (including a few fields zoffline never gave it, so seemingly those are unused defaults, e.g. the 1970 date string) along with a couple other fields, but it's not related to this issue since originally the Zwift client just performed a GET on that url and this bug was present then.

Glad you mentioned it though since it made me look at it and see that the client is posting back the uncorrupted segment results given to it. That rules out a segment corruption issue, which was becoming my first choice for this bug.

**Comment by oldnapalm** [comment]
Isn't world_time rounded to nearest minute? (int(time.time()//60*60)*1000)

Just a guess because when showing segment results, the smallest unit Zwift uses is minutes (X minutes/hours/days/weeks/months ago).

**Comment by oldnapalm** [comment]
Experimenting with https://github.com/jeroni7100/zwift-packet-monitor I noticed it displays world_time as a long with a negative low part, e.g.

`worldTime: Long { low: -1252083071, high: 40, unsigned: false },`

Creating a `new Long(-1252083071, 40, false)` in JavaScript it converts to 174841576065

This number is related to unix epoch (at least for the last few hours), the difference is 1414016075 seconds (considering both in milliseconds).

Tried to insert that negative low part in segment_results (changing world_time to int64) and Zwift displays the date correctly (only in Alpe segments, other segments still show as "just now").

![81307174-c2529280-9056-11ea-8d26-db3bdcfdb172](https://user-images.githubusercontent.com/38410858/81317701-8c67db00-9063-11ea-8052-17f2d59e1dc2.png)

**Comment by oldnapalm** [comment]
Experimenting with https://github.com/jeroni7100/zwift-packet-monitor I noticed it shows a value for worldTime which is related to real time (at least for the last few hours), the difference is around 1414016075 seconds (considering both in milliseconds). Will check again in a few days.

```
const ZwiftPacketMonitor = require('@zwfthcks/zwift-packet-monitor')

const monitor = new ZwiftPacketMonitor('en0')

monitor.on('incomingPlayerState', (playerState, serverWorldTime) => {
  console.log(Date.now() - playerState.worldTime.toNumber());
})

monitor.start()
```

```
1414016074747
1414016074660
1414016074781
1414016074700
1414016074629
1414016074698
1414016074654
1414016074667
1414016077968
1414016074678
1414016074742
1414016074678
1414016074837
1414016074626
1414016074679
1414016074717
1414016074613
1414016074685
1414016074566
1414016074647
1414016074719
1414016074562
1414016074577
```

**Comment by oldnapalm** [comment]
> edit 3: UDP node is indeed the segment result timing issue. Need to figure out wtf world time is though before resolving.

Which parts of the UDP node do you think would be necessary to get this working?

Regarding world_time, maybe it's too early but I think this is a good approximation https://github.com/oldnapalm/zwift-offline/commit/c0711513ee8c8af4692bd52ce3a96dc170a75308

That kind of makes sense, since the first beta came out in September 2014. It could be milliseconds since October 2014.

**Comment by oldnapalm** [comment]
The world_time difference is still holding.

Apparently the client doesn't need to connect to the UDP server to show segment results correctly.

Client sends hello to TCP telemetry server which returns a list of UDP servers. Tried to return the same data to the client and the segments look right.

```
        if bytearray(msg) == bytearray.fromhex('000C0010E80718006800C72457F8'):
            con.sendall(bytearray.fromhex('008210e8071800c201370a17080110061a0e35342e3134392e3131302e32313820ce170a16080010001a0d31382e3233362e3234302e383120ce17100a181e2003ca01400a1d080110061a17080110061a0e35342e3134392e3131302e32313820ce170a1c080010001a16080010001a0d31382e3233362e3234302e383120ce1710ce17'))
        else:
            con.sendall(bytearray.fromhex('000710e80718005801'))
```

![Captura de Tela 2020-05-11 às 08 32 28](https://user-images.githubusercontent.com/38410858/81562323-07363c00-936b-11ea-8e76-bf4b51b82e33.png)

It doesn't actually connect to the UDP server, just keeps trying

```
NETCLIENT:[INFO] TCP host 127.0.0.1:3023
NETCLIENT:[INFO] Connecting to TCP server...
NETCLIENT:[INFO] Saying hello to TCP server (largest wa is 0)
NETCLIENT:[INFO] UDP host 18.236.240.81:3022
NETCLIENT:[INFO] Connecting to UDP server...
NETCLIENT:[INFO] UDP host 54.149.110.218:3022
NETCLIENT:[INFO] Connecting to UDP server...
NETCLIENT:[WARN] UDP connection timeout (1 so far), reconnection attempt 1
NETCLIENT:[INFO] Connecting to UDP server...
NETCLIENT:[WARN] UDP connection timeout (2 so far), reconnection attempt 2
NETCLIENT:[INFO] Connecting to UDP server...
NETCLIENT:[WARN] UDP connection timeout (3 so far), reconnection attempt 3
NETCLIENT:[INFO] Connecting to UDP server...
```

**Comment by zoffline** [comment]
Awesome work! Thanks!
